### DISCUSSION TITLE
Running `MPFS_ICICLE_KIT_REFERENCE_DESIGN.tcl` causes Libero to close silently.

### DISCUSSION THREAD

**Discussion by Geekdude** [polarfire-soc/polarfire-soc-documentation#170] [Reference designs, FPGA and development kits] [answered]
I am using Windows 10 Enterprise and Libero 2022.1.0.10. 

When I checkout the master branch, open Libero, and execute the `MPFS_ICICLE_KIT_REFERENCE_DESIGN.tcl` script, libero executes the script for a few minutes and then closes without giving a script report. When I open the project afterward, the project seems incomplete, which leads me to believe, that the script caused libero to crash silently. 

Please help me debug why libero is closing and to successfully generate the reference design.

Thanks,
Aaron Young

**Comment by hughbreslin** [comment, marked as answer]
Hi @Geekdude 

I'm sorry thats happening for you. Could you try generating the design in [batch mode](https://github.com/polarfire-soc/icicle-kit-reference-design#generating-designs-from-the-command-line---no-gui) and seeing if that succeeds? 

The script executes successfully on my end on my Windows 10 machine and on Ubuntu but I have seen occasional crashes when I run it, usually a restart resolves the issue. One thing to check is the path length to the folder you have the script in, do you have a long path to the script? Sometimes you can be stung by the path length limitation on Windows. 

If all else fails I can generate a copy of the project for you here and share it so you can open the design.

**Reply by Geekdude** [reply]
Hi @hughbreslin,

Thank you for your reply. I am glad to hear that you are able to build on Windows and Ubuntu since those are the two OSs I am trying to build on. However, I have not had success with either OS.

## Windows 10
- I tried running in batch mode as you suggested. The prompt returns immediately, but it seems to be running in the background. Is there a way to run the script in the foreground or to get an indication of when the build is complete?
- After waiting until it has completed, I get this output log. [MPFS_ICICLE_KIT_REFERENCE_DESIGN.log](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9284171/MPFS_ICICLE_KIT_REFERENCE_DESIGN.log)
- The path length should be OK since I have the repo checked out as `D:\icicle-kit-reference-design`.
- The build fails with
    ```
    Error: The ENVM configuration has the following errors:
            Client 'BOOT_MODE_1_ENVM_CLIENT': Unable to open memory file 'D:/icicle-kit-reference-design/script_support/hss-bm1-p0.hex'.
    Error: The command 'configure_envm' failed.
    Error: Failure when executing Tcl script. [ Line 279 ]
    Error: The Execute Script command failed.
    The MPFS_ICICLE_I2C_LOOPBACK project was closed.
    ````

## Ubuntu
- The build fails with this output: 
[run.log](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9284191/run.log)
- This output fails with:
    ```
    Job: "premap" terminated with error status: 2
    See log file: "/home/7ry/Data/1_Projects/sawtooth/icicle-kit-reference-design/MPFS_ICICLE_I2C_LOOPBACK/synthesis/synlog/MPFS_ICICLE_KIT_BASE_DESIGN_premap.srr"
    ```
- `MPFS_ICICLE_I2C_LOOPBACK/synthesis/synlog/MPFS_ICICLE_KIT_BASE_DESIGN_premap.srr` Contains: 
[MPFS_ICICLE_KIT_BASE_DESIGN_premap.srr.log](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9284211/MPFS_ICICLE_KIT_BASE_DESIGN_premap.srr.log)

**Reply by hughbreslin** [reply]
Ok cool so it looks like its generating the design on both OS's. 

# Windows

Re running the script in the foreground, unfortunately as far as I know theres no way to do this.

It looks like you passed the "HSS_UPDATE" argument and maybe "EXPORT_FPE" or "GENERATE_BITSTREAM"? "HSS_UPDATE" needs `wget` set up to work - the issue seems to be that it can't find the HSS hex file which could be down to the download not working. Can you see the hex file in the `script_support` directory?

To test the design is generating you could run the script with no arguments and see that it passes generating the design you could then open the GUI and modify the design and run the flow.

# Ubuntu

Again the design seems to have generated here and the issue you're seeing is with the synthesis tool. As the design is the same on both Windows and Linux and its synthesizing on Windows I don't think the issue is the design itself. Could you be missing a package for Synplify to run? Are you able to synthesize other designs or is it just this one that has an issue?

**Reply by Geekdude** [reply]
Thank you again for your reply; it was very helpful for debugging the issue. 

I switched to CentOS 7.9, and I am now able to generate the bitstream successfully. Now I am running into trouble when I try programming the FPGA. I get the same error when I try via the GUI and via batch mode. Please let me know if you have any advice to solve this error.

Batch mode programming output:
```
===============================================================
Programmer: Embedded FP6 '4CC43AFA' : Version = EF9B
===============================================================
Device not found
programmer '4CC43AFA' : Embedded FlashPro6
===============================================================
Programmer: Embedded FP6 '4CC43AFA' : Version = EF9B
===============================================================
Device not found
programmer '4CC43AFA' : Embedded FlashPro6
programmer '4CC43AFA' : Scan Chain...
Error:  programmer '4CC43AFA' : Signal Integrity Failure
        Integrity Check Pattern Not Found.
        Integrity Check Pattern :
        550FAAF000FF0000FFFF
        IrScan Error.
        TDO stuck at 1
        Chain Analysis Failed.
Error:  programmer '4CC43AFA' : Data Bit length : 3920
Error:  programmer '4CC43AFA' : Compare Data  : FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
Error:  programmer '4CC43AFA' : Scan Chain FAILED.
Error:  The command 'run_tool' failed.
Error:  Failure when executing Tcl script. [ Line 289 ]
Error:  The Execute Script command failed.
The MPFS_ICICLE_I2C_LOOPBACK project was closed.
Exported log file MPFS_ICICLE_KIT_REFERENCE_DESIGN.log.
```

GUI Programming Error:
![image](https://user-images.githubusercontent.com/2541055/184780650-f67070c5-5996-4f26-951c-850ecacbca10.png)

**Reply by hughbreslin** [reply]
No problem, glad I could help 🙂 

On the programming issue I suspect this is down to jumper settings on the kit, to use the eFP6 you need to close J9 (its just below the external FlashPro header - check out this [image](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/boards/mpfs-icicle-kit-es/updating-icicle-kit/updating-icicle-kit-design-and-linux.md#jumper-settings)). I can create the same error by programming with J9 open when using the eFP6.

**Reply by TommyMurphyTM1234** [reply]
> * I tried running in batch mode as you suggested. The prompt returns immediately, but it seems to be running in the background. Is there a way to run the script in the foreground or to get an indication of when the build is complete?

An indirect way to do that is to enable logging when running the script with the Libero command line option `LOGFILE:output.log` (obviously you can name the log file whatever you want) 

https://onlinedocs.microchip.com/pr/GUID-3D4C6B83-916E-4A35-8894-A0E1E63ED7CF-en-US-2/index.html?GUID-EC46CBCA-23C3-4F55-9E91-93F2FA742DBA

and then use `tail -f output.log` to monitor progress and watch for completion.

https://man7.org/linux/man-pages/man1/tail.1.html

On Windows, if there is no Unix/Linux style `tail` utility available (e.g. via CygWin or similar) then you can probably use the Windows PowerShell to run Libero in logging enabled scripted mode and use the PowerShell `Get-Content` command and its `-Tail` option to do the same thing.

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/get-content?view=powershell-7.2

`LOGFILE:CON` may work in a regular Windows command shell to direct Libero logging to the console (via the Windows/DOS reserved `CON` filename) obviating the need for `tail`, but I haven't tried it.

https://superuser.com/questions/86999/why-cant-i-name-a-folder-or-file-con-in-windows

**Reply by Geekdude** [reply]
@TommyMurphyTM1234 Thank you for the suggestion. Logging to a file and watching the file is how I have been checking to see when the execution is complete. Still not as convenient as the script running in the foreground (Canceling a run is more difficult.)

@hughbreslin Thanks again for the help. The missing jumper was indeed why I was having the programming error.  Another challenge I ran into was caused by running the development tools in a VM. When I programmed the board, the USB device would disconnect and would not get reattached to the VM, which resulted in a programming error. I resolved this error by setting up [Libvirt automatic USB passthrough scripts](https://www.raddinox.com/libvirt-usb-passthrough-scripts)[^1]. Now I am able to build and load the reference design and program the eMMC memory successfully.

[^1]: I had to modify the hotplug script slightly for CentOS. See [my fork](https://github.com/Geekdude/usb-libvirt-hotplug) for the changes.
